### ISSUE TITLE
Decide which libraries to use

### ISSUE THREAD

**Issue by dionhaefner** [DHI-GRAS/itsybitsy#1]
@j08lue @mrpgraae:

I have now added [4 different implementations of the same scraping algorithm](https://github.com/DHI-GRAS/itsybitsy/tree/master/itsybitsy/spider) to explore what we can use in `goget` and similar future projects. I am a bit on the fence about which one to go for, so I would be happy to hear your opinions. The candidates are:

#### 1. Explicitly multi-threaded + Requests

Pro:
- Universally compatible
- Fastest implementation for large trees

Con:
- Most complicated code
- High memory usage when using 100+ threads (about 2GB of memory for 10k pages)
- Prone to race conditions and possibly deadlocks
- Some overhead for smaller pages

#### 2. Gevent greenlets + monkey patched Requests

Pro:
- Universally compatible
- Quite readable code
- Good memory usage, no race conditions

Con:
- Slowest implementation by far (barely faster than sequential)
- Monkey patching feels hacky

#### 3. Asyncio + aiohttp

Pro:
- Most elegant solution
- Minimal memory usage
- No threading overhead

Con:
- Only supports Python 3.5 and above
- aiohttp feels not as robust as requests yet
- aiohttp seems to download whole data files sometimes even though we just want their headers, which can cause this implementation to become quite slow (and download **a ton** of data)

#### 4. Asyncio + Requests in concurrent.futures threads

Alright, stay with me. To use Requests with asyncio, you can make it non-blocking by putting all I/O tasks into a separate thread managed by `concurrent.futures`.

Pro:
- Basically the same clean code as no. 3
- No speed penalty from aiohttp
- One interface fits all

Con:
- Only supports Python 3.5 and above
- Some thread overhead for smaller jobs
- Memory consumption as high as multithreaded implementation

**Comment by j08lue** [comment]
2. is out. If the speedup is marginal, it does not justify the fancy code.

3. sounds quite bad with huge download volume.

4. has cleaner code than 1. but is only Python 3.5 (which I find is acceptable - it's not, like, 2016 anymore), but also the same high memory consumption - which is the smallest evil of the ones you listed, I guess.

**Comment by dionhaefner** [comment]
I managed to fix the issue with `aiohttp` (it was actually [this bug](https://github.com/aio-libs/aiohttp/issues/1821)). Now, implementation 3 is the fastest of all 😄

**Comment by dionhaefner** [comment]
Looks like we got a winner! 🎊

Congrats to number 3.
